tests/integration/servicecatalog/cmd_service_test.go

```diff
@@ -416,4 +416,31 @@ var _ = Describe("odo service command tests", func() {
 			Expect(stdOut).To(ContainSubstring("DB_HOST"))
 		})
 	})
+
+	Context("When the application is deleted", func() {
+		JustBeforeEach(func() {
+			preSetup()
+			app = helper.RandString(6)
+		})
+		JustAfterEach(func() {
+			cleanPreSetup()
+		})
+		It("should delete the service(s) in the application as well", func() {
+			helper.CmdShouldPass("odo", "service", "create", "--app", app, "-w", "dh-postgresql-apb", "--project", project, "--plan", "dev",
+				"-p", "postgresql_user=luke", "-p", "postgresql_password=secret",
+				"-p", "postgresql_database=my_data", "-p", "postgresql_version=9.6")
+			ocArgs := []string{"get", "serviceinstance", "-o", "name", "-n", project}
+			helper.WaitForCmdOut("oc", ocArgs, 1, true, func(output string) bool {
+				return strings.Contains(output, "dh-postgresql-apb")
+			})
+
+			helper.CmdShouldPass("odo", "app", "delete", app, "-f")
+			// This timeout is needed because it takes some time to delete the
+			// application and the components/services that belong to it on the
+			// cluster
+			time.Sleep(60 * time.Second)
```

Comment:
Why not ping the server constantly and check if the service list is deleted or not? This will pause the execution for 60 secs even if the deletion occurred after 30 secs.

Reply:
That's what I did in this commit https://github.com/openshift/odo/pull/2491/commits/efe3eb0f6edcbe041311c697b36253ed091b49dd.

But that command expects the command to exit successfully. It's a wrapper on top of `CmdShouldPass` whereas in this case, we are expecting `odo service list` command to fail.
Hence I went with this approach.

Reply:
Why not create a wrapper on top of `CmdShouldFail`? Or use the use the oc command above and keep checking till the output doesn't contain `dh-postgresql-apb`?

Reply:
> Why not create a wrapper on top of CmdShouldFail? 

That's doable but I didn't do it since it's just one check. Had I come across more such scenarios, it would have been more worth adding it, IMO.

> Or use the use the oc command above and keep checking till the output doesn't contain dh-postgresql-apb?

That's what I was checking for earlier. But it exits with non-zero return code and causes the test to fail since it's a wrapper on top of `CmdShouldPass`.

Reply:
> That's what I was checking for earlier. But it exits with non-zero return code and causes the test to fail since it's a wrapper on top of CmdShouldPass.

I think that shouldn't use CmdShouldPass or CmdShouldFail. It should just get the ouput of the command. @amitkrout Can you please have a look?

Reply:
@mik-dass Thanks for pointing me. will have a look for sure

Reply:
@amitkrout do you have any comments on this? I'd really like this PR to be able to move forward even if that means having to make some changes from my side.

Reply:
Sorry, i was fully packed up with other test stuff.

@dharmit I would suggest lets not encourage ```sleep()``` in our test code unless it is unavoidable.

Can you please explore if you can change the test code as per @mik-dass suggestion ?

Reply:
While I agree with what @mik-dass has suggested, if we do it in this PR, I will have to make changes at all those places in the code where `helper.WaitForCmdOut` is being called:
```sh
$ grep -r "helper.WaitForCmdOut" tests
tests/integration/servicecatalog/cmd_link_unlink_test.go:                       helper.WaitForCmdOut("oc", ocArgs, 1, true, func(output string) bool {
tests/integration/servicecatalog/cmd_link_unlink_test.go:                       helper.WaitForCmdOut("oc", ocArgs, 1, true, func(output string) bool {
tests/integration/servicecatalog/cmd_link_unlink_test.go:                       helper.WaitForCmdOut("oc", ocArgs, 1, true, func(output string) bool {
tests/integration/servicecatalog/cmd_service_test.go:                   helper.WaitForCmdOut("oc", ocArgs, 1, true, func(output string) bool {
tests/integration/servicecatalog/cmd_service_test.go:                   helper.WaitForCmdOut("oc", ocArgs, 1, true, func(output string) bool {
tests/integration/servicecatalog/cmd_service_test.go:                   helper.WaitForCmdOut("odo", odoArgs, 1, true, func(output string) bool {
tests/integration/servicecatalog/cmd_service_test.go:                   helper.WaitForCmdOut("oc", ocArgs, 1, true, func(output string) bool {
tests/integration/servicecatalog/cmd_service_test.go:                   helper.WaitForCmdOut("oc", ocArgs, 1, true, func(output string) bool {
tests/integration/servicecatalog/cmd_service_test.go:                   helper.WaitForCmdOut("oc", ocArgs, 1, true, func(output string) bool {
tests/integration/servicecatalog/cmd_service_test.go:                   helper.WaitForCmdOut("oc", ocArgs, 1, true, func(output string) bool {
tests/integration/servicecatalog/cmd_service_test.go:                   helper.WaitForCmdOut("oc", ocArgs, 1, true, func(output string) bool {
tests/integration/servicecatalog/cmd_service_test.go:                   helper.WaitForCmdOut("oc", ocArgs, 1, true, func(output string) bool {
tests/integration/servicecatalog/cmd_service_test.go:                   helper.WaitForCmdOut("oc", ocArgs, 1, true, func(output string) bool {
```
I'd rather open a new issue based on this conversation which can then be picked up and fixed!

Reply:
@dharmit You can, However your test script itself is not completed till you are not handling the sleep(). IMO this not a good test practice to write unhealthy script in the current test change and fix it in the next followup pr. 

Then the better approach would be w.r.t. the situation you are facing - First fix the dependency in a separate pr and merge. Then rebase the changes to this pr. WDYT ?

Reply:
Okay. Let's open a new issue based on this discussion, have team's thoughts on it and open a PR for it. Till that new issue is fixed, we can mark this PR as blocked on that issue. Does that sound OK?

Reply:
opened a PR which adds a test helper that does what you have mentioned @dharmit https://github.com/openshift/odo/pull/2676

Reply:
Why not just use `helper.WaitForCmdOut` with `oc` command to check for `dh-postgresql-apb` service?

Just PR is stale for 2 months just because of this one thing?

Reply:
> Why not just use helper.WaitForCmdOut with oc command to check for dh-postgresql-apb service?
oc get serviceinstance

Because, IIRC (since it's 2 months), `WaitForCmdOut` doesn't work if the cmd fails. We're trying to execute a command that's expected to fail.

> Is this PR stale for 2 months just because of this one thing?

That's my bad. I got into OperatorHub piece and left my sight from this one. Just discussed on a call. I'll address it this week.

Reply:
> Because, IIRC (since it's 2 months), `WaitForCmdOut` doesn't work if the cmd fails. We're trying to execute a command that's expected to fail.

Using `oc get serviceinstance` would work with `WaitForCmdOut`

Reply:
I've used `oc get serviceinstance` with `WaitForCmdOut` and pushed the changes. When I tried to run this test locally, it would not succeed in spite of having found the sub-string that we're looking for in the output of the command. I've pushed it however to get feedback from the CI.